### ISSUE TITLE
Prerelease 0.2.0rc1 no longer returns log probs with llama.cpp

### ISSUE THREAD

**Issue by oj-sec** [guidance-ai/guidance#1064]
**The bug**
Updating from guidance==0.1.16 to prerelease guidance==0.2.0rc1 causes model.log_probs() to return 0 rather than the true log probs for a generation when using the llama.cpp backend. I have tested GGUF quants of models based on Llama, Mistral and Gemma and observed this behaviour to be model agnostic.

**To Reproduce**
Reproduction Colab notebook [here ](https://gist.github.com/oj-sec/1a34b653811aa8b2857178087775fc76) - involves uninstalling and reinstalling Guidance, but the change in output between installs is:
```python
# With guidance==0.1.16
from guidance import models, gen, select
import math

llm = models.LlamaCpp(f"./models/{model}", n_gpu_layers=40, n_ctx=2000, compute_log_probs=True)
output = llm + "You flip a coin. The result is: " + gen(name="coinflip", regex="(heads|tails)")
logprobs = output.log_prob("coinflip")
prob = round(math.exp(logprobs), 5)
print(f"Output:{output['coinflip']}\nLP: {logprobs}\nP: {prob}")
```
```
You flip a coin. The result is: heads
Output:heads
LP: -1.1534752799652015
P: 0.31554
```
```python
# With guidance==0.2.0rc1
from guidance import models, gen, select
import math

llm = models.LlamaCpp(f"./models/{model}", n_gpu_layers=40, n_ctx=2000, compute_log_probs=True)
output = llm + "You flip a coin. The result is: " + gen(name="coinflip", regex="(heads|tails)")
logprobs = output.log_prob("coinflip")
prob = round(math.exp(logprobs), 5)
print(f"Output:{output['coinflip']}\nLP: {logprobs}\nP: {prob}")
```
```
You flip a coin. The result is: heads
Output:heads
LP: 0.0
P: 1.0
```

**System info (please complete the following information):**
 - OS (e.g. Ubuntu, Windows 11, Mac OS, etc.): Google Colab, Ubuntu 22.04
 - Guidance Version (`guidance.__version__`): 0.2.0rc1 (both from pypi and from https://github.com/microsoft/guidance.git@77fc3999e1545c10f17e6da1b6cbd1feeaa1ca1a)
 
 If I can provide any further info please let me know. Huge thanks for this amazing library.

**Comment by hudson-ai** [comment]
Hi @oj-sec thanks for bringing this up, and our apologies if it's impacting your workflow. The new parser that we're using in `0.2.0rc1` is considerably *faster* than what was running in previous versions, but it currently has a few limitations that we need to continue working on (probability outputs is probably the primary one). So, it's on our radar.

Thank you for submitting the issue!

**Comment by woitee** [comment]
@hudson-ai Since 0.2.0 has released and the visualization seems to display probabilities, can we expect this issue to get fixed (soon)?

**Comment by hudson-ai** [comment]
Hi @woitee, sorry for the late reply. We're currently displaying probabilities for outputted **tokens**, but we don't yet have a satisfactory solution for mapping these back to probabilities for a given capture (which is a string that may or may not align to token boundaries). Getting this working again is on the roadmap, but I don't yet have a timeline for you.

Would you mind giving me an idea of how you like to use this feature in practice? More example usages may help motivate a solution :)

**Comment by woitee** [comment]
@hudson-ai Thanks for the reply! The main thing I need this for is categorization, I use guidance for classifying e-mails and free-text forms into categories, and I used to provide users with information on how "certain" AI is in these, based on probabilities. If it worked only for `select` captures, it would solve 90% of usage for me. But I assume that still has to deal with token boundaries correctly and is far from trivial.

Can I access the token probabilities? I might be able to arrange my systems such that a token definitely ends before the select clause and different tokens are at the starts of options - and create some work arounds :)

**Comment by ya0guang** [comment]
Seems like `log_prob` also returns 0 for transformers model, but I'm not sure if this is expected.

**Comment by oj-sec** [comment]
For anyone who is still wanting to access this information, it's available under a the `_trace_nodes` attribute attached to responses. 

A function to return the generated token probability and the top_k at that poisiton as a dict is:

```python
from guidance.trace import TokenOutput

def extract_token_probs(trace_nodes):
    """
    Function to extract token probabilities from a
    guidance output._trace_nodes set.
    """
    results = []

    for node in trace_nodes:
        for output in node.output:
            if isinstance(output, TokenOutput):
                result = {
                    "id": node.identifier,
                    "token": output.token.token,
                    "prob": output.token.prob,
                    "top_k": []
                }

                if output.top_k:
                    for alt in output.top_k:
                        result["top_k"].append({
                            "token": alt.token,
                            "prob": alt.prob,
                        })

                results.append(result)

    return results
```

**Comment by woitee** [comment]
@oj-sec I'm getting "nan" in all the prob fields. Is there any setting that needs to be toggled?

I'm currently using Meta-Llama-3-8B-Instruct.Q5_K_M.gguf.

**Comment by woitee** [comment]
After reading 0.2.5 release notes, setting echo=True did the trick. Not sure if that is the "correct" solution.

**Comment by ErenAta16** [comment]
Confirmed -- this isn't a coincidence, it's how it's wired. In both `models/_llama_cpp.py` and `models/_transformers.py`, `echo` is passed straight through as `enable_token_probabilities=echo` (and `enable_top_k=echo`) to the engine. In `_engine.py`, whenever `enable_token_probabilities` is false, `prob` defaults to `float('nan')` for fast-forwarded tokens. So `echo=True` genuinely is the intended way to get probabilities today -- it's just that `echo` is doing double duty as "enable live rendering" and "enable probability computation," since both require the extra softmax-over-full-logits work. There's no separate flag to get probs without also implicitly turning on echo-driven rendering machinery.
